src/emsarray/operations/cache.py

```diff
@@ -0,0 +1,108 @@
+"""
+Operations for making caching keys for given datasets and attribute dicts.
+"""
+import hashlib
+import marshal
+
+import numpy
+import xarray
+
+import emsarray
+
+
+def hash_attributes(hash: "hashlib._Hash", attributes: dict) -> None:
+    """
+    Updates the provided hash with with a marshal serialised byte representation of the given attribute dictionary.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: dict
+        Expects a marshal compatible dictionary
+    """
+    # Prepend the marshal encoding version
+    marshal_version = 4
+    hash_int(hash, marshal_version)
+    # Specify marshal encoding version when serialising
+    attribute_dict_marshal_bytes = marshal.dumps(attributes, marshal_version)
+    # Prepend the number of attributes
+    hash_int(hash, len(attributes))
+    # Prepend the size of the pickled attributes
+    hash_int(hash, len(attribute_dict_marshal_bytes))
+    hash.update(attribute_dict_marshal_bytes)
+
+
+def hash_string(hash: "hashlib._Hash", value: str) -> None:
+    """
+    Updates the provided hash with with a utf-8 encoded byte representation of the provided string.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: str
+        Expects a string that can be encoded in utf-8
+    """
+    # Prepend the str length
+    hash_int(hash, len(value))
+    hash.update(value.encode('utf-8'))
+
+
+def hash_int(hash: "hashlib._Hash", value: int) -> None:
+    """
+    Updates the provided hash with an encoded byte representation of the provided int.
+
+    Parameters
+    ----------
+    hash : hashlib-style hash instance
+        The hash instance to update with the given attribute dict.
+        This must follow the interface defined in :mod:`hashlib`.
+    attributes: int
+        Expects an int that can be represented in a numpy int32
+    """
+    # Prepend the int bit length
+    hash.update(numpy.int32(value.bit_length()).tobytes())
+    hash.update(numpy.int32(value).tobytes())
```

Comment:
An int32 value will always be four bytes, so prepending the bit length is superfluous. We only need to prepend the size of variable length things like strings and arrays.

This does potentially leave us vulnerable to overflows if someone hashes something longer than 2^32 bits long, but that can be handled another day...